### ISSUE TITLE
multiple search criteria on the same MDD field

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#742]
**Reported by @cristianbogdan on 26 Apr 2010 16:18 UTC**
currently it is not possible to put more search criteria on the same MDD field.

there will be a form field name collision. this can be solved in a similar way with multiple forms.

Migrated-From: http://trac.makumba.org/ticket/1197

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 28 Apr 2010 11:45 UTC**
It seems that I predicted such a problem when implementing, http://makumba.svn.sourceforge.net/viewvc/makumba/trunk/makumba/classes/org/makumba/forms/tags/CriterionTag.java?revision=4487&view=markup#l_66 reads:

// FIXME: we can't have the same field as the primary criterion yet => it would lead to two inputs with
// the same name. 
// The fix should be to give a different name to the input, and store a mapping input-name => field
// name, so we can construct the query correctly.
// For now, we just prevent errors by checking that this field was not yet used in this search form
if (getForm().containsInput(getInputName())) {
    throw new ProgrammerError(
            "This search form already contains one critireon tag with the same field as the first criterion. For now, this is not supported in Makumba!");
}


However, I just made a test page (http://makumba.svn.sourceforge.net/viewvc/makumba/trunk/makumba/webapps/tests/forms-oql/testMakSearchFormRepeatedCriterion.jsp?revision=4893, or http://localhost:8080/tests/forms-oql/testMakSearchFormRepeatedCriterion.jsp in the Makumba tomcat environment).

When commenting out the exception throwing above, it seems to work like a charm.

As assumed by me initially, the query condition composing, and thus the resultList, have no issues at all (they don't mind where the input data comes from).
Surprisingly, however, also the mak:searchField/mak:input deals with correctly setting the selected options after a form submission.

Cristi, can you try this uncommenting mentioned above in your application, and see if it works?

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 2 May 2010 16:15 UTC**
(In reply to comment #1)
> As assumed by me initially, the query condition composing, and thus the
> resultList, have no issues at all (they don't mind where the input data comes
> from).

i think they do, see below

> Cristi, can you try this uncommenting mentioned above in your application, and
> see if it works?

sorry it took a while as i had to understand the TLD phenomenon first

after commenting out the code you indicate, the page does run but the two fields have the same form/HTML name. also the generated WHERE does not honor the two different criteria. it is:

EXISTS (FROM o.parti o_parti_set WHERE o_parti_set IN SET ($parti))

instead of 

EXISTS (FROM o.parti o_parti_set WHERE o_parti_set IN SET ($parti)) 
AND  
EXISTS (FROM o.parti o_parti_set1 WHERE o_parti_set1 IN SET ($parti1))

basically due  the two criteria are "united" in the same, this is equivalent with an OR in the condition above, which is why apparently your test page works (btw, how to run those? i had to stop a test battery in mid-air with a breakpoint to have some proper language data). 

but the semantic of my form (and i guess the semantic of most search forms) is AND.  

speaking of which, i looked at the doc for search forms but i didn't find anything about OR vs AND between fields. is there anything like that?

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 4 May 2010 13:59 UTC**
(In reply to comment #2)
> after commenting out the code you indicate, the page does run but the two
> fields have the same form/HTML name. also the generated WHERE does not honor
> the two different criteria. it is:
> 
> EXISTS (FROM o.parti o_parti_set WHERE o_parti_set IN SET ($parti))
> 
> instead of 
> 
> EXISTS (FROM o.parti o_parti_set WHERE o_parti_set IN SET ($parti)) 
> AND  
> EXISTS (FROM o.parti o_parti_set1 WHERE o_parti_set1 IN SET ($parti1))


Ok, so finally I think I understand what you want.. 
But I am not sure if I understand why you actually need it this way. Can you please detail a bit your data structure, and describe the semantics of the fields?

My main question is that if you split the values of a certain field among some (predefined?) criterion, wouldn't it then make sense to actually have those values in separate fields?

> basically due  the two criteria are "united" in the same, this is equivalent
> with an OR in the condition above, which is why apparently your test page works
> (btw, how to run those? i had to stop a test battery in mid-air with a
> breakpoint to have some proper language data). 
> 
> but the semantic of my form (and i guess the semantic of most search forms) is
> AND.  
> 
> speaking of which, i looked at the doc for search forms but i didn't find
> anything about OR vs AND between fields. is there anything like that?


Each condition is automatically AND combined; there was an initial idea to allow modifying the whole search form to OR, that could be easily done with an attribute in the searchForm tag.


If you want to combine some conditions with "OR", then you can only do that by adding them in one criterion, with fields="name, surname, shortname".
Then you get
where conditionX AND (name=$y OR surname=$y OR shortname=$y) AND conditionZ

More complex combinations aren't supported.. But I'd like to see a use case first anyhow..

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 4 May 2010 17:27 UTC**
(In reply to comment #3)
> But I am not sure if I understand why you actually need it this way. Can you
> please detail a bit your data structure, and describe the semantics of the
> fields?

well initially there was one field and one search box. like any search form. then the customer said that they wanted to differentiate between 3 companies (who belong to the customer) and all other companies 

the data structure is a contract, which has contractual parts. in the contractual parts, the 3 companies are mixed with all other companies.  most contracts are between one company of the customer and some other company.

in order to have easier searching, they choose the company that belongs to them in one list box and the company that doesn't in another. incidentally they are on the same set.

> My main question is that if you split the values of a certain field among some
> (predefined?) criterion, wouldn't it then make sense to actually have those
> values in separate fields?

yes you may argue that this is wrong modeling but my argument is that search forms would only benefit from this. as you see,  the MQL that should be generated is not much different from the case when there would be 2 separate fields.

but if it's hard to do, i can probably move the 3 things in another set, though fixing the data for that may take some time. i need migration tools :)

yet i am still blocked in using search forms in this app by bug 1166/bug 1167

/* 
not important for this bug: 
> If you want to combine some conditions with "OR", then you can only do that by
> adding them in one criterion, with fields="name, surname, shortname".

interesting, i didn't find any doc for this. all examples i saw in the doc are with one field per criterion. 
*/

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:24 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
